pkg/operator/controller/certificate/publish_certs.go

```diff
@@ -0,0 +1,160 @@
+package certificate
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"reflect"
+
+	ingressv1alpha1 "github.com/openshift/cluster-ingress-operator/pkg/apis/ingress/v1alpha1"
+	"github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ensureRouterCertsGlobalSecret will create, update, or delete the global
+// certificates secret as appropriate.
+func (r *reconciler) ensureRouterCertsGlobalSecret(secrets []corev1.Secret, ingresses []ingressv1alpha1.ClusterIngress) error {
+	desired, err := desiredRouterCertsGlobalSecret(secrets, ingresses)
+	if err != nil {
+		return err
+	}
+	current, err := r.currentRouterCertsGlobalSecret()
+	if err != nil {
+		return err
+	}
+	switch {
+	case desired == nil && current == nil:
+		// Nothing to do.
+	case desired == nil && current != nil:
+		if deleted, err := r.deleteRouterCertsGlobalSecret(current); err != nil {
+			return fmt.Errorf("failed to ensure router certificates secret was unpublished: %v", err)
+		} else if deleted {
+			r.recorder.Eventf(current, "Normal", "UnpublishedRouterCertificates", "Unpublished router certificates")
+		}
+	case desired != nil && current == nil:
+		if created, err := r.createRouterCertsGlobalSecret(desired); err != nil {
+			return fmt.Errorf("failed to ensure router certificates secret was published: %v", err)
+		} else if created {
+			r.recorder.Eventf(desired, "Normal", "PublishedRouterCertificates", "Published router certificates")
+		}
+	case desired != nil && current != nil:
+		if updated, err := r.updateRouterCertsGlobalSecret(current, desired); err != nil {
+			return fmt.Errorf("failed to update published router certificates secret: %v", err)
+		} else if updated {
+			r.recorder.Eventf(desired, "Normal", "UpdatedPublishedRouterCertificates", "Updated the published router certificates")
+		}
+	}
+	return nil
+}
+
+// desiredRouterCertsGlobalSecret returns the desired router-certs global
+// secret.
+func desiredRouterCertsGlobalSecret(secrets []corev1.Secret, ingresses []ingressv1alpha1.ClusterIngress) (*corev1.Secret, error) {
+	if len(ingresses) == 0 || len(secrets) == 0 {
+		return nil, nil
+	}
+
+	nameToSecret := map[string]*corev1.Secret{}
+	for i, certSecret := range secrets {
+		nameToSecret[certSecret.Name] = &secrets[i]
+	}
+
+	ingressToSecret := map[*ingressv1alpha1.ClusterIngress]*corev1.Secret{}
+	for i, ingress := range ingresses {
+		secretName := fmt.Sprintf("router-certs-%s", ingress.Name)
```

Comment:
This is duplicated from `desiredRouterDeployment()` — should it be extracted? reminds me of `enforceEffectiveDomain`, `haTypeForInfra`, etc. in `controller.go`.

Reply:
Are you playing mind games? Looks like you already extracted it into `RouterDefaultCertificateSecretName`.

Reply:
Changed to use `RouterDefaultCertificateSecretName`.  Thanks!

Reply:
Hmm, I was actually referring to the [effective value calculation](https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator/controller/controller_router_deployment.go#L202)

```go
secretName := fmt.Sprintf("router-certs-%s", ci.Name)
if ci.Spec.DefaultCertificateSecret != nil && len(*ci.Spec.DefaultCertificateSecret) > 0 {
  secretName = *ci.Spec.DefaultCertificateSecret
}
```

Reply:
Can address in a followup

Reply:
I'll get it in a separate PR so as not to delay this feature.

Reply:
https://github.com/openshift/cluster-ingress-operator/pull/149